src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala

```diff
@@ -0,0 +1,298 @@
+package io.iohk.scalanet.peergroup
+
+import java.net.InetSocketAddress
+import java.security.PrivateKey
+import java.security.cert.X509Certificate
+
+import io.iohk.decco.BufferInstantiator.global.HeapByteBuffer
+import io.iohk.decco._
+import io.iohk.scalanet.peergroup.PeerGroup.TerminalPeerGroup
+import io.iohk.scalanet.peergroup.TLSPeerGroup._
+import io.netty.bootstrap.{Bootstrap, ServerBootstrap}
+import io.netty.buffer.{ByteBuf, Unpooled}
+import io.netty.channel._
+import io.netty.channel.nio.NioEventLoopGroup
+import io.netty.channel.socket.SocketChannel
+import io.netty.channel.socket.nio.{NioServerSocketChannel, NioSocketChannel}
+import io.netty.handler.codec.bytes.ByteArrayEncoder
+import io.netty.handler.codec.{LengthFieldBasedFrameDecoder, LengthFieldPrepender}
+import io.netty.handler.ssl.{SslContextBuilder, SslHandshakeCompletionEvent}
+import io.netty.util
+import monix.eval.Task
+import monix.reactive.Observable
+import monix.reactive.subjects.{PublishSubject, ReplaySubject, Subject}
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Promise
+import scala.util.Success
+
+/**
+  * PeerGroup implementation on top of TCP.
+  * FIXME currently this class makes use of netty's LengthFieldPrepender to perform packet reassembly. This means
+  * the encoded bytes provided by the callers codec are not identical to the bytes put on the wire (since a
+  * length field is prepended to the byte stream). This class therefore cannot be used to talk to general services
+  * that are not instances of TCPPeerGroup.
+  *
+  * @param config bind address etc. See the companion object.
+  * @param codec a decco codec for reading writing messages to NIO ByteBuffer.
+  * @tparam M the message type.
+  */
+class TLSPeerGroup[M](val config: Config)(implicit codec: Codec[M]) extends TerminalPeerGroup[InetMultiAddress, M]() {
+
+  private val log = LoggerFactory.getLogger(getClass)
+
+  private[scalanet] class ServerChannelImpl[M](val nettyChannel: SocketChannel)(
+      implicit codec: Codec[M]
+  ) extends Channel[InetMultiAddress, M] {
+
+    private val log = LoggerFactory.getLogger(getClass)
+    private val messageSubject = ReplaySubject[M]()
```

Comment:
Be aware that anyone subscribing to this subject will receive the whole message history instead of the ones produced after subscribing to the subject, if that's the intended behavior, please add a comment explaining why.

Reply:
This an ongoing issue. We had a test failure (about a month ago) where the new channel event was not delivered to subscribers until _after_ the first message had been delivered to that channel's message stream. This meant that the subscriber missed the start of the message stream.

The only suggestion to avoid this problem was to replay the message stream.

Nonetheless, this replay approach is potentially problematic is production scenarios because of the obvious memory problems associated with replay buffering. We need a confirmed and robust solution that can be rolled out to peer group implementations. Obviously we don't want to block scalanet development in the meantime so replaying is there for the time being.

Reply:
Ok, please create a ticket, otherwise, we may forget.